### PR TITLE
Retry on ciliumnode update/create

### PR DESCRIPTION
**Pull request by ashrayjain**
Currently, we don't retry on updates/creates for the
CiliumNode when the agent starts up.

This is suboptimal because the operator and the agent
can end up "fighting" each other since both of them create/update
the CiliumNode resource. This can cause the agent to continually
fail to start and prolong the time before a node is ready for use.

Signed-off-by: Ashray Jain <ashrayj@palantir.com>

```release-note
Retry on conflicts when creating/updating CiliumNode objects on agent startup
```

### CODE REVIEW

**Comment by ungureanuvladvictor** [comment]
test-me-please

**Comment by ungureanuvladvictor** [comment]
test-me-please

File: pkg/nodediscovery/nodediscovery.go [hunk 1]
```diff
@@ -218,17 +221,47 @@ func (n *NodeDiscovery) UpdateCiliumNodeResource() {
...
 
-	performUpdate := true
-	nodeResource, err := ciliumClient.CiliumV2().CiliumNodes().Get(context.TODO(), nodeTypes.GetName(), metav1.GetOptions{})
-	if err != nil {
-		performUpdate = false
-		nodeResource = &ciliumv2.CiliumNode{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: nodeTypes.GetName(),
-			},
+	for retryCount := 0; retryCount < maxRetryCount; retryCount++ {
```
Comment (aanm) [inline pkg/nodediscovery/nodediscovery.go:223]: Thank you for the PR. This was on my TODO list as well. I think we don't need to rely in retries. We could use watchers instead. Something like this: https://github.com/kubernetes/kubernetes/pull/88935/commits/e053fdd08ae00d1bff3dd42a2c034f4810f8a55d#diff-934760ea1a4e195e49bcd10625cf4e46R384

Does it makes sense?

**Review by aanm — CHANGES_REQUESTED** [review]

**Review by aanm — CHANGES_REQUESTED** [review]
Changes look good to me. only one small nit:
```diff
diff --git a/pkg/nodediscovery/nodediscovery.go b/pkg/nodediscovery/nodediscovery.go
index e52cb1c6c..cb4b4dce6 100644
--- a/pkg/nodediscovery/nodediscovery.go
+++ b/pkg/nodediscovery/nodediscovery.go
@@ -16,7 +16,6 @@ package nodediscovery
 
 import (
        "context"
-       "k8s.io/apimachinery/pkg/api/errors"
        "strings"
        "time"
 
@@ -42,8 +41,9 @@ import (
        nodeTypes "github.com/cilium/cilium/pkg/node/types"
        "github.com/cilium/cilium/pkg/option"
        "github.com/cilium/cilium/pkg/source"
-
        cnitypes "github.com/cilium/cilium/plugins/cilium-cni/types"
+
+       "k8s.io/apimachinery/pkg/api/errors"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
```

**Review by aanm — APPROVED** [review]

**Pull request merged**